src/main/cpp/src/datetime_parser.cu

```diff
@@ -0,0 +1,560 @@
+      } else {
+        if (i < segments_len && (':' == b || ' ' == b)) {
+          if (!is_valid_digits(i, current_segment_digits)) {
+            return thrust::make_tuple(error_us, false);
+          }
+          segments[i]            = current_segment_value;
+          current_segment_value  = 0;
+          current_segment_digits = 0;
+          i += 1;
+        } else {
+          return thrust::make_tuple(error_us, false);
+        }
+      }
+    } else {
+      if (6 == i) { digits_milli += 1; }
+      // We will truncate the nanosecond part if there are more than 6 digits, which results
+      // in loss of precision
+      if (6 != i || current_segment_digits < 6) {
+        current_segment_value = current_segment_value * 10 + parsed_value;
+      }
+      current_segment_digits += 1;
+    }
+    j += 1;
+  }
+
+  if (!is_valid_digits(i, current_segment_digits)) { return thrust::make_tuple(error_us, false); }
+  segments[i] = current_segment_value;
+
+  while (digits_milli < 6) {
+    segments[6] *= 10;
+    digits_milli += 1;
+  }
+
+  cudf::string_view time_zone;
+  if (tz.has_value()) {
+    time_zone = tz.value();
+  } else {
+    time_zone = cudf::string_view(default_time_zone, default_time_zone_char_len);
+  }
+
+  segments[0] *= year_sign.value_or(1);
+  // above is ported from Spark.
+
+  // set components
+  auto components = timestamp_components{segments[0],
+                                         static_cast<int8_t>(segments[1]),
+                                         static_cast<int8_t>(segments[2]),
+                                         static_cast<int8_t>(segments[3]),
+                                         static_cast<int8_t>(segments[4]),
+                                         static_cast<int8_t>(segments[5]),
+                                         segments[6]};
+  if (default_time_zone_char_len == 0) {
+    // invoke from `string_to_timestamp_without_time_zone`
+    if (just_time || !allow_time_zone && tz.has_value()) {
+      return thrust::make_tuple(error_us, false);
+    } else {
+      return to_utc_timestamp(components);
+    }
+  } else {
+    // invoke from `string_to_timestamp`
+    if (just_time) {
+      // Update here to support the following format:
+      //   `[h]h:[m]m:[s]s.[ms][ms][ms][us][us][us][zone_id]`
+      //   `T[h]h:[m]m:[s]s.[ms][ms][ms][us][us][us][zone_id]`
+      // by set local date in a time zone: year-month-day.
+      // Above 2 formats are time zone related, Spark uses LocalDate.now(zoneId)
+
+      // do not support currently
+      return thrust::make_tuple(error_us, false);
+    } else {
+      return to_utc_timestamp(components, time_zone);
+    }
+  }
+}
+
+struct parse_timestamp_string_fn {
+  cudf::column_device_view const d_strings;
+  const char* default_time_zone;
+  cudf::size_type default_time_zone_char_len;
+  bool allow_time_zone;
+  bool allow_special_expressions;
+  // TODO the following should be passed in.
```

Comment:
Is there an issue number for this TODO so we don't forget about it? Also because this is likely to impact the APIs, it feels like a fast/simple fix to do now rather than wait for a follow on. It would also allow us to actually test that we got the right answer.

Reply:
This comment is invalid now:
Refer to: 
https://github.com/NVIDIA/spark-rapids-jni/pull/1539#issuecomment-1888842389